planner/core/rule_join_reorder.go

```diff
@@ -165,7 +170,34 @@ func (s *joinReOrderSolver) optimizeRecursive(ctx sessionctx.Context, p LogicalP
 				break
 			}
 		}
-		if len(curJoinGroup) > ctx.GetSessionVars().TiDBOptJoinReorderThreshold || !isSupportDP {
+
+		baseGroupSolver := &baseSingleGroupJoinOrderSolver{
+			ctx:        ctx,
+			otherConds: otherConds,
+			eqEdges:    eqEdges,
+			joinTypes:  joinTypes,
+		}
+
+		joinGroupNum := len(curJoinGroup)
+		useGreedy := joinGroupNum > ctx.GetSessionVars().TiDBOptJoinReorderThreshold || !isSupportDP
+		if hasOuterJoin {
+			ctx.GetSessionVars().StmtCtx.AppendWarning(ErrInternal.GenWithStack("leading hint is inapplicable when we have outer join"))
+		} else {
+			if hintInfo != nil && hintInfo.leadingJoinOrder != nil {
```

Comment:
Should we check `hintInfo != nil && leadingJoinOrder != nil` before `hasOuterJoin` otherwise it'll generate a warning even if there is no leading hint?

Reply:
Yes, you are right. I'll fix it later.